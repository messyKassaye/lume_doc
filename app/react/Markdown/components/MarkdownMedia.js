import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactPlayer from 'react-player';
import { Icon } from 'UI';

const propsToConfig = props => {
  const config = { url: '', options: {} };

  let parsedProps = props.config.replace(/\(|\)/g, '').split(',');
  config.url = parsedProps.shift();

  parsedProps = (parsedProps.join(',') || '{}').replace(/&quot;/g, '"');

  try {
    parsedProps = JSON.parse(parsedProps);
  } catch (error) {
    parsedProps = {};
  }

  config.options = parsedProps;

  return config;
};

class MarkdownMedia extends Component {
  timeLinks(_timelinks) {
    const timelinks = _timelinks || {};
    return Object.keys(timelinks).map((timeKey, index) => {
      const seconds = timeKey
        .split(':')
        .reverse()
        .reduce((_seconds, n, _index) => _seconds + parseInt(n, 10) * 60 ** _index, 0);
      return (
        <div className="timelink" key={index} onClick={this.seekTo.bind(this, seconds)}>
          <b>
            <Icon icon="play" /> {timeKey}
          </b>
          <span>{timelinks[timeKey]}</span>
        </div>
      );
    });
  }

  seekTo(seconds) {
    this.player.seekTo(seconds);
  }

  render() {
    const config = propsToConfig(this.props);
    const { compact } = this.props;
    const dimensions = { width: '100%' };
    if (compact) {
      dimensions.height = '100%';
    }
    return (
      <div className={`video-container ${compact ? 'compact' : ''}`}>
        <div>
          <ReactPlayer
            className="react-player"
            ref={ref => {
              this.player = ref;
            }}
            url={config.url}
            {...dimensions}
            controls
          />
        </div>
        <div>{this.timeLinks(config.options.timelinks)}</div>
      </div>
    );
  }
}

MarkdownMedia.defaultProps = {
  compact: false,
};

MarkdownMedia.propTypes = {
  compact: PropTypes.bool,
  config: PropTypes.string.isRequired,
};

export default MarkdownMedia;
