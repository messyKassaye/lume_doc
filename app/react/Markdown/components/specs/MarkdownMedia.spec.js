import React from 'react';
import { shallow } from 'enzyme';
import ReactPlayer from 'react-player';

import MarkdownMedia from '../MarkdownMedia';

describe('MarkdownMedia', () => {
  let component;
  let instance;
  let props;

  beforeEach(() => {
    props = {
      config:
        '(https://www.vimeo.com/253530307, {"timelinks": {"02:10": "A rude awakening", "05:30": "Finally, you are up!"}})',
    };
  });

  const render = () => {
    component = shallow(<MarkdownMedia {...props} />);
    instance = component.instance();
  };

  describe('render', () => {
    it('should render an iframe with the correct video id', () => {
      render();
      expect(component).toMatchSnapshot();
    });

    it('should use compact class name if compact prop is set', () => {
      props.compact = true;
      render();
      expect(component).toMatchSnapshot();
    });

    it('should assign the ReactPlayer ref to this.player', () => {
      const playerRef = {};
      render();

      const ReactPlayerComponent = component.find(ReactPlayer);
      ReactPlayerComponent.getElement().ref(playerRef);
      expect(instance.player).toBe(playerRef);
    });

    describe('Video timeline', () => {
      beforeEach(() => {
        render();
        instance.player = { seekTo: jasmine.createSpy('seekTo') };
      });

      it('should render the timelinks', () => {
        const links = component.find('.timelink');
        expect(links.length).toBe(2);
      });

      it('should interact with the player', () => {
        const firstTimeLink = component.find('.timelink').at(0);
        const secondTimeLink = component.find('.timelink').at(1);

        firstTimeLink.simulate('click');
        expect(instance.player.seekTo).toHaveBeenCalledWith(2 * 60 + 10);
        secondTimeLink.simulate('click');
        expect(instance.player.seekTo).toHaveBeenCalledWith(5 * 60 + 30);
      });
    });
  });
});
