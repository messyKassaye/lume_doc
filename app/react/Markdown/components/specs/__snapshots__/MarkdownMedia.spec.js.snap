// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownMedia render should render an iframe with the correct video id 1`] = `
<div
  className="video-container "
>
  <div>
    <ReactPlayer
      className="react-player"
      config={
        Object {
          "dailymotion": Object {
            "params": Object {
              "api": 1,
              "endscreen-enable": false,
            },
            "preload": false,
          },
          "facebook": Object {
            "appId": "1309697205772819",
          },
          "file": Object {
            "attributes": Object {},
            "dashVersion": "2.9.2",
            "forceAudio": false,
            "forceDASH": false,
            "forceHLS": false,
            "forceVideo": false,
            "hlsOptions": Object {},
            "hlsVersion": "0.10.1",
            "tracks": Array [],
          },
          "mixcloud": Object {
            "options": Object {
              "hide_cover": 1,
            },
          },
          "soundcloud": Object {
            "options": Object {
              "buying": false,
              "download": false,
              "liking": false,
              "sharing": false,
              "show_comments": false,
              "show_playcount": false,
              "visual": true,
            },
          },
          "twitch": Object {
            "options": Object {},
          },
          "vimeo": Object {
            "playerOptions": Object {
              "autopause": false,
              "byline": false,
              "portrait": false,
              "title": false,
            },
            "preload": false,
          },
          "wistia": Object {
            "options": Object {},
          },
          "youtube": Object {
            "embedOptions": Object {},
            "playerVars": Object {
              "iv_load_policy": 3,
              "modestbranding": 1,
              "playsinline": 1,
              "rel": 0,
              "showinfo": 0,
            },
            "preload": false,
          },
        }
      }
      controls={true}
      height="360px"
      light={false}
      loop={false}
      muted={false}
      onBuffer={[Function]}
      onBufferEnd={[Function]}
      onDisablePIP={[Function]}
      onDuration={[Function]}
      onEnablePIP={[Function]}
      onEnded={[Function]}
      onError={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      onProgress={[Function]}
      onReady={[Function]}
      onSeek={[Function]}
      onStart={[Function]}
      pip={false}
      playbackRate={1}
      playing={false}
      playsinline={false}
      progressInterval={1000}
      style={Object {}}
      url="https://www.vimeo.com/253530307"
      volume={null}
      width="100%"
      wrapper="div"
    />
  </div>
  <div>
    <div
      className="timelink"
      key="0"
      onClick={[Function]}
    >
      <b>
        <Connect(Icon)
          icon="play"
        />
         
        02:10
      </b>
      <span>
        A rude awakening
      </span>
    </div>
    <div
      className="timelink"
      key="1"
      onClick={[Function]}
    >
      <b>
        <Connect(Icon)
          icon="play"
        />
         
        05:30
      </b>
      <span>
        Finally, you are up!
      </span>
    </div>
  </div>
</div>
`;

exports[`MarkdownMedia render should use compact class name if compact prop is set 1`] = `
<div
  className="video-container compact"
>
  <div>
    <ReactPlayer
      className="react-player"
      config={
        Object {
          "dailymotion": Object {
            "params": Object {
              "api": 1,
              "endscreen-enable": false,
            },
            "preload": false,
          },
          "facebook": Object {
            "appId": "1309697205772819",
          },
          "file": Object {
            "attributes": Object {},
            "dashVersion": "2.9.2",
            "forceAudio": false,
            "forceDASH": false,
            "forceHLS": false,
            "forceVideo": false,
            "hlsOptions": Object {},
            "hlsVersion": "0.10.1",
            "tracks": Array [],
          },
          "mixcloud": Object {
            "options": Object {
              "hide_cover": 1,
            },
          },
          "soundcloud": Object {
            "options": Object {
              "buying": false,
              "download": false,
              "liking": false,
              "sharing": false,
              "show_comments": false,
              "show_playcount": false,
              "visual": true,
            },
          },
          "twitch": Object {
            "options": Object {},
          },
          "vimeo": Object {
            "playerOptions": Object {
              "autopause": false,
              "byline": false,
              "portrait": false,
              "title": false,
            },
            "preload": false,
          },
          "wistia": Object {
            "options": Object {},
          },
          "youtube": Object {
            "embedOptions": Object {},
            "playerVars": Object {
              "iv_load_policy": 3,
              "modestbranding": 1,
              "playsinline": 1,
              "rel": 0,
              "showinfo": 0,
            },
            "preload": false,
          },
        }
      }
      controls={true}
      height="100%"
      light={false}
      loop={false}
      muted={false}
      onBuffer={[Function]}
      onBufferEnd={[Function]}
      onDisablePIP={[Function]}
      onDuration={[Function]}
      onEnablePIP={[Function]}
      onEnded={[Function]}
      onError={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      onProgress={[Function]}
      onReady={[Function]}
      onSeek={[Function]}
      onStart={[Function]}
      pip={false}
      playbackRate={1}
      playing={false}
      playsinline={false}
      progressInterval={1000}
      style={Object {}}
      url="https://www.vimeo.com/253530307"
      volume={null}
      width="100%"
      wrapper="div"
    />
  </div>
  <div>
    <div
      className="timelink"
      key="0"
      onClick={[Function]}
    >
      <b>
        <Connect(Icon)
          icon="play"
        />
         
        02:10
      </b>
      <span>
        A rude awakening
      </span>
    </div>
    <div
      className="timelink"
      key="1"
      onClick={[Function]}
    >
      <b>
        <Connect(Icon)
          icon="play"
        />
         
        05:30
      </b>
      <span>
        Finally, you are up!
      </span>
    </div>
  </div>
</div>
`;
